### DISCUSSION TITLE
Detect the difference between left/right clicks on the MenuBarExtra item

### DISCUSSION THREAD

**Discussion by leonschrijvers** [orchetect/MenuBarExtraAccess#2] [Ideas]
Hi,

Thanks for your awesome work! I have a question which may be out of scope of this project, but because in-depth information related to MenuBarExtra is still hard to find, I'm giving it a shot :)

Would it be possible to detect the difference between left and right clicks on the MenuBarExtra item? I have used this in the past in applications that used NSStatusBar with a statusItem, see here for some background information on how to get that to work: https://stackoverflow.com/questions/59635971/show-nsmenu-only-on-nsstatusbarbutton-right-click 

I was hoping that this could also be done with MenuBarExtra natively-ish, without having to fallback on the AppDelegate/NSStatusBar approach. What's your take on this? Thanks!

**Comment by orchetect** [comment]
MenuBarExtra is managing that and only connects left (primary) click.

From my experience thus far, it may be possible by accessing the NSStatusItem and manipulating it. But there was a reason I didn’t outwardly expose the NSStatusItem in this library. MenuBarExtra operates under the assumption that it owns and manages the status item. Which means we can’t anticipate what Apple may do in future iterations of SwiftUI. So a hacky solution that works now could break or crash in the future.

It’s a balancing act. Apple needs to add 1st-party support for these features ideally (including presentation binding, left/right mouse differentiation, etc.). So submitting radars for feature requests to Apple and having patience may produce some results eventually. This library is just meant to act as a bridge to coax more functionality out of MenuBarExtra because, ever the optimist, I want to adopt new technologies and show my support for them.

**Reply by leonschrijvers** [reply]
Thanks for your reply! I believe we have the same ideas about adopting new technologies, that's the reason I wanted to see if left/right mouse button differentiation could be added to MenuBarExtra in some way. I understand that the following the official way (submitting feature requests) will eventually lead to better results, but in the meantime I was hoping to find a way that prevents me from going back to the AppDelegate/NSStatusBar approach.

You said that it may be possible by accessing the NSStatusItem and manipulating it. Could you give some pointers on how this may be done? In the MenuBarExtraAccess code base, I see some NSStatusItem references already, but it's hard for me to see how to proceed from there. 

By the way, if this is a topic that you don't want to spend any time on, I totally understand. Thanks!

**Reply by orchetect** [reply]
It's a fair question and I've considered making direct access to the NSStatusItem available. Let me have another dig and see if it feels viable to do so.

**Reply by orchetect** [reply]
I've added access to the NSStatusItem in the scene modifier. It can be mutated in the closure, or stored to a local state var for later access. Pushed to `main` branch.

```swift
@main struct MyApp: App {
    @State var isStatusMenuPresented: Bool = false
    @State var menuBarStatusItem: NSStatusItem?

    var body: some Scene {
        MenuBarExtra("MyApp Menu", systemImage: "0.circle.fill") {
            // ... menu contents ...
        }
        .menuBarExtraAccess(isPresented: $isStatusMenuPresented) { statusItem in
            // can do one-time setup of NSStatusItem here
            statusItem.button?.appearsDisabled = true
            
            // or if access to the NSStatusItem is needed later, it may be stored in a local state var
            menuBarStatusItem = statusItem
        }
    }
}
```

This doesn't address the left/right click issue. That's potentially a much bigger problem. Thing is, MenuBarExtra sets itself up as a standard button with left click only and there's a good chance that if you subclass any of the views it uses in order to implement right-click handling, it could break or crash MenuBarExtra's internals. Also, MenuBarExtraAccess relies on the button property of the `NSStatusItem` being a valid `NSStatusBarButton` to operate correctly.

I don't have time to pursue the right-click issue at the moment, but if you'd like to do your own exploration and let me know if anything works that could be useful. If there is a reliable way to implement right-click detection it would be a good thing to know and could either be included as an Example project or possibly a feature of the library.

**Reply by leonschrijvers** [reply]
Thanks for the update! I have been fiddling around with the click detection for some time now, but I cannot get it to work. I tried to set the action property of the status item button to a custom method to process the click events, but the method never gets called. This may be of the concerns that you already had, but it may also be the case that I made a mistake, so I will post the code anyway:

```swift
@main
struct MenuBarTestApp: App {
    @State var isStatusMenuPresented: Bool = false
    
    let statusItemHelper: StatusItemHelper = StatusItemHelper()
    
    var body: some Scene {
        MenuBarExtra("MyApp Menu", systemImage: "0.circle.fill") {
            ContentView()
        }
        .menuBarExtraAccess(isPresented: $isStatusMenuPresented) { statusItem in
            print("menuBarExtraAccess")
            
            statusItem.button?.action = #selector(statusItemHelper.statusItemButtonClicked(sender:))
            statusItem.button?.target = statusItemHelper
            statusItem.button?.sendAction(on: [.leftMouseUp, .rightMouseUp])
        }
    }
}

class StatusItemHelper {
    @objc func statusItemButtonClicked(sender: NSStatusBarButton) {
        print("statusItemButtonClicked")
        let event = NSApp.currentEvent!
        
        if event.type ==  NSEvent.EventType.rightMouseUp {
            print("Right click!")
        } else {
            print("Left click!")
        }
    }
}
```

Because the selector wants a method marked with @objc (which cannot be done inside a struct), I created a helper class for this. But as said before, the method never gets called. Do you any ideas on why I could try to get this to work?

**Reply by orchetect** [reply]
The action is only sent on primary button clicks (as far as I remember). So that's why you aren't getting any feedback for right clicks.

From my sleuthing on Google a while ago I recall seeing that you have to subclass either the status item's `NSView` or the `NSStatusBarButton` and override the `func rightMouseDown(with event: [NSEvent])` method. That's why I mentioned the risks of subclassing because it's uncharted waters with regards to the black box of how MenuBarExtra currently interacts with the status item (or how it could interact in future).

**Reply by leonschrijvers** [reply]
First of all, to clarify my previous message: the event also didn't fire for left clicks, so that's why I assumed that something was wrong with how I wired things up.

However, I tried your suggestion of subclassing NSView and that works! By overriding mouseDown and rightMouseDown, you have full control over both events. Some observations:

- To disable opening/closing the MenuBarExtra upon click, comment the `super.moseDown()` or `super.rightMouseDown()` calls. This may be interesting in combination with other MenuBarExtraAccess functionality
- If you first call the `super.mouseDown()` or `super.rightMouseDown()` event and then call your own closure, then the closure only gets called after closing the MenuBarExtra, which may be unexpected behavior. Switching the order (as shown in the code below) fixes this.
- The events only get called when opening the MenuBarExtra (by using either left mouse button or right mouse button).
- Subclassing `NSStatusBarButton` is also possible, but then the MenuBarExtra also gets a 'Button' text, which can be removed by setting the `mouseHandlerView.title` to an empty string. The MenuBarExtra also doesn't open in that case, so have to do that yourself. 
- It's not sure to me if subclassing `NSStatusBarButton` has any other pros or cons compared to `NSView`.

Thanks for your support, and I hope this helps other people as well to get more out of MenuBarExtra. There are still some things I want to improve, but this is definitely progress! The working code:


The working code:

```swift
@main
struct MenuBarTestApp: App {
    @State var isStatusMenuPresented: Bool = false
        
    var body: some Scene {
        MenuBarExtra("MyApp Menu", systemImage: "0.circle.fill") {
            ContentView()
        }
        .menuBarExtraAccess(isPresented: $isStatusMenuPresented) { statusItem in
            if let button = statusItem.button {
                let mouseHandlerView = MouseHandlerView(frame: button.frame)
                
                mouseHandlerView.onMouseDown = {
                    print("leftMouseDown")
                }
                
                mouseHandlerView.onRightMouseDown = {
                    print("rightMouseDown")
                }
                                
                button.addSubview(mouseHandlerView)
            }
        }
    }
}

class MouseHandlerView: NSView {
    var onRightMouseDown: (() -> Void)? = nil
    var onMouseDown: (() -> Void)? = nil

    override func rightMouseDown(with event: NSEvent) {
        if onRightMouseDown != nil {
            onRightMouseDown!()
        }
        
        super.rightMouseDown(with: event)
    }
    
    override func mouseDown(with event: NSEvent) {
        if onMouseDown != nil {
            onMouseDown!()
        }
        
        super.mouseDown(with: event)
    }
}
```

**Reply by orchetect** [reply]
Interesting, thanks so much for the detailed info.

Using `addSubview` on the button means that the existing underlying view and button is still there (which in this case is good since it interferes the least with MenuBarExtra's own interaction with the view). That is probably the safest way to go to avoid potential unknowns with how Apple may alter MBE's behavior in future.

My gut tells me that the `mouseDown` override shouldn't be used at all in the custom view. That view should only be for right-click handling. The left click is already handled by MenuBarExtra itself. I'd consider this hack as purely a way to add right-click functionality.

Also, this is a better approach to the right-click override:

```swift
override func rightMouseDown(with event: NSEvent) {
    if let onRightMouseDown {
        onRightMouseDown()
    } else {
        super.rightMouseDown(with: event)
    }
}
```

You don't want to pass the event to super if you're handling it. Otherwise you have both your handler and super's handler being called which is not what you want.

**Reply by leonschrijvers** [reply]
Thanks for the improved implementation, that's a lot cleaner. 

Your explanation makes a lot of sense, and I can see that only overriding `rightMouseDown` is the safest approach. However for the use case I am interested in, I want to intercept the left click and run some own code and open the MBE with the right click. For sake of completeness, the final code to make that happen:

```swift
@main
struct MenuBarTestApp: App {
    @State var isStatusMenuPresented: Bool = false
        
    var body: some Scene {
        MenuBarExtra("MyApp Menu", systemImage: "0.circle.fill") {
            ContentView()
        }
        .menuBarExtraAccess(isPresented: $isStatusMenuPresented) { statusItem in
            if let button = statusItem.button {
                let mouseHandlerView = MouseHandlerView(frame: button.frame)
                
                mouseHandlerView.onMouseDown = {
                    print("leftMouseDown")
                }
                                
                button.addSubview(mouseHandlerView)
            }
        }
    }
}

class MouseHandlerView: NSView {
    var onRightMouseDown: (() -> Void)? = nil
    var onMouseDown: (() -> Void)? = nil

    override func rightMouseDown(with event: NSEvent) {
        if let onRightMouseDown {
            onRightMouseDown()
        } else {
            super.rightMouseDown(with: event)
        }
    }
    
    override func mouseDown(with event: NSEvent) {
        if let onMouseDown {
            onMouseDown()
        } else {
            super.mouseDown(with: event)
        }
    }
}
``` 

Thanks for your help!

**Reply by orchetect** [reply]
Gotcha 👍 Great that it seems to be working.

I'm on the fence about adding this as a library feature, as bit by bit we're kind of rewriting MenuBarExtra which is a slippery slope.

It may be useful as a demo project at least. If you feel like throwing one together (maybe using the latest platform demo as a template and gut it first) and PR it you're welcome to, but no pressure. I can throw it together later too.

**Reply by leonschrijvers** [reply]
Sure, I can see if I can throw something together. Is it an idea to rename the Demo folder to Demos and create a subfolder for the current Demo, so that Demos can host multiple demo projects?

**Reply by orchetect** [reply]
Right, yes. I was thinking of my MacControlCenterUI library which already had multiple demos.
